dashboard/app/models/levels/blockly.rb

```diff
@@ -639,6 +643,38 @@ def localized_blocks_with_placeholder_texts(blocks)
     block_xml.serialize(save_with: XML_OPTIONS).strip
   end
 
+  # Localizing placeholder texts in all possible block types within markdown.
+  # @param markdown [String]
+  # @return [String]
+  # @see unit test for an example of blocks that contain placeholder texts.
+  def localized_markdown_with_placeholder_texts(markdown)
+    return if markdown.nil?
+
+    placeholders = Hash.new
+    processed_markdown = Nokogiri::HTML(markdown, &:noblanks)
+
+    processed_markdown.xpath("//block[@type=\"text\"]").each do |block|
+      title = block.at_xpath("./title[@name=\"TEXT\"]")
+      next unless title&.content&.present?
+
+      text_key = Digest::MD5.hexdigest title.content
+      localized_text = I18n.t(
+        text_key,
+        scope: [:data, :placeholder_texts, name],
+        default: nil,
+        smart: true
+      )
+      placeholders[title.content] = localized_text if localized_text
+    end
+
+    # Replace each string
+    placeholders.keys.each do |original_str|
+      markdown = markdown.gsub(original_str, placeholders[original_str])
+    end
```

Comment:
Could this potentially replace parts of the string **outside** the placeholder block?

Reply:
For visibility:
The localization of markdown is handled in this [new shared function now](https://github.com/code-dot-org/code-dot-org/pull/43575/files#diff-d628235b9ae09147729a8305879b909985a10f740f7c90a49ef9a749a082128cR473). Which should address this concern.